File: 10-adapterArray/go/challenge/partTwo.go
package challenge

import (
	"sort"
	"strconv"
	"strings"
)

func PartTwo(instr string) int {
	var joltages []int
	for _, x := range strings.Split(strings.TrimSpace(instr), "\n") {
		i, err := strconv.Atoi(x)
		if err != nil {
			panic(err)
		}
		joltages = append(joltages, i)
	}
	sort.Ints(joltages)

	routeLengths := map[int]int{0: 1}

	for _, joltage := range joltages {
		var totalRoutes int
		for _, n := range []int{1, 2, 3} {
			v, exists := routeLengths[joltage-n]
			if !exists {
				v = 0
			}
			totalRoutes += v
		}
		routeLengths[joltage] = totalRoutes
	}

	return routeLengths[joltages[len(joltages)-1]]
}

File: 10-adapterArray/python/partTwo.py
def partTwo(instr: str) -> int:
    joltages = sorted([int(x) for x in instr.strip().split("\n")])

    route_lengths = {0: 1}
    for joltage in joltages:
        total_routes = 0
        # Get the route lengths for the three previous joltages
        for n in [1, 2, 3]:
            total_routes += route_lengths.get(joltage - n, 0) 
        print(joltage, total_routes)
        route_lengths[joltage] = total_routes

    return route_lengths[max(joltages)]
